フォローによるタイムライン変更を実装(DAIKI)

i-k-i-a-d opened the pull request:
・フォローしている人とフォロワーを一覧で表示するbladeファイルを作成。
・新しく作った2つのbladeファイルを表示させるためのルーティングとUserController内のメソッドをそれぞれ作成。
・タブをクリックするとそれぞれのタブがアクティブになるように三項演算子で条件分岐作成。
・フォロー、フォロー解除後のリダイレクト処理をback();関数を使って修正。
・ログイン者が自分のユーザー詳細画面の中のフォロー中のタブを開くとフォローを解除するボタンが表示されるように実装。
・ログイン者が自分のフォロワータブを開き、もし、フォロワーをログイン者がフォローしていないときは、フォローするボタンが表示されるように実装。

ベストプラクティスかは自身ありませんが、とりあえず動くものには仕上がってます。
レビューよろしくお願いします。

i-k-i-a-d committed:
フォローしている人とフォロワーを一覧で表示するbladeファイルを作成。
新しく作った2つno
2つのbladeファイルを表示させるためのルーティングとUserController内のメソッドwo
をそれぞれ作成。
タブをクリックするとそれぞれのタブがアクティブになるように三項演算子で条件分岐作成。
フォロー・フォロー解除後のリダイレクト処理をback();関数を使って修正。

Osk0830 commented on resources/views/users/detail.blade.php:
```diff
@@ -32,9 +32,9 @@
-            <li class="nav-item"><a href="" class="nav-link {{ Request::is() ? 'active' : '' }}">タイムライン</a></li>
-            <li class="nav-item"><a href="#" class="nav-link">フォロー中</a></li>
-            <li class="nav-item"><a href="#" class="nav-link">フォロワー</a></li>
+            <li class="nav-item"><a href="" class="nav-link {{ Request::route()->named('users.show') ? 'active' : '' }}">タイムライン</a></li>
```
- これでも間違いないではないですが、下記記事②の `Request::routeIs()`を用いるともう少しだけ処理がスマートになり記述も短くなりますので、今回の場合はこちらの方がいいかもですね。

https://shiro-changelife.com/laravel-url-judge/

i-k-i-a-d replied:
そんな記述方法があったとは知りませんでした。
勉強になります。

Osk0830 commented on resources/views/users/followed.blade.php:
```diff
@@ -0,0 +1,62 @@
+        <ul class="nav nav-tabs nav-justified mb-3">
+            <li class="nav-item"><a href="{{ route('users.show', $user->id) }}" class="nav-link">タイムライン</a></li>
+            <li class="nav-item"><a href="{{ route('users.following', $user->id) }}" class="nav-link">フォロー中</a></li>
+            <li class="nav-item"><a href="" class="nav-link {{ Request::route()->named('users.followed') ? 'active' : '' }}">フォロワー</a></li>
```
- こちらについては前述の通りです。

i-k-i-a-d replied:
修正しました。

Osk0830 commented on resources/views/users/followed.blade.php:
```diff
@@ -0,0 +1,62 @@
+            <li class="nav-item"><a href="" class="nav-link {{ Request::route()->named('users.followed') ? 'active' : '' }}">フォロワー</a></li>
+        </ul>
+        <ul class="list-unstyled">
+            @foreach($user->followers()->get() as $follow)
```
- ここでデータを取得することも可能ですが、今後フォローとフォロワーの共通部分を部品化することを考えると、フォローの場合とフォロワーの場合とでBladeの記述を最大限同じにしておくことが望ましく、BladeごとにBlade内で異なるメソッドを実行しなければならない状況は色々と不便さを助長します。今回は、コントローラーの各アクション内でデータを取得してBladeに渡す形にしておきましょう。

i-k-i-a-d replied:
部品化の意味ようやく理解しました。
この作業が部品化という呼ばれ方をするのはエンジニア共通認識でしょうか？？

Osk0830 replied:
部品化、パーツ化と言ったり、コンポーネント化と言うことが多い気がします。

Osk0830 commented on resources/views/users/followed.blade.php:
```diff
@@ -0,0 +1,62 @@
+                        <img class="mr-2 rounded-circle" src="{{ Gravatar::src($follow->email, 55) }}" alt="ユーザのアバター画像">
+                        <p class="mt-3 mb-0 d-inline-block"><a href="">{{ $follow->name }}</a></p>
+                    </div>
+                    <div class="">
+                        @if (Auth::id() === $user->id && !Auth::user()->followCheck($follow->id))
+                            <div class="d-flex justify-content-between w-75 pb-3 m-auto">
+                                <form method="POST" action="{{ route('follow', $follow->id) }}">
+                                    @csrf
+                                    <button type="submit" class="btn btn-primary">フォローする</button>
+                                </form>
+                            </div>
+                        @endif
+                    </div>
```
- <div class=""></div>がif文の外にあるため、条件に合致しなかった際に不要な空<div>ができてしまわないでしょうか？
- ボタン表示の理屈と内容は現状でOKだと思います。ただ、今後部品化を図っていくことになると、少し変更を加える必要が出てきそうですが、そちらについては次のタスクで考えてみるようにしましょう。

i-k-i-a-d replied:
完全に見落としでした。
htmlでの記述場所も意識しておかないといけないですね。

Osk0830 reviewed (changes requested):
お疲れ様でした。
全体的によくできていると思います。
少しだけコメントしましたので、確認よろしくお願いします。

i-k-i-a-d committed:
Request::routeIs()を用いてタブをアクティブにする際の三項演算子の条件式を修正。
フォローデータをUserController内のメソッドで取得してbladeファイルへ引き継ぎするよう修正。
if文をdivタグの外に出し条件式に合わなかった際に余計なスペースが入らないようにした。
フォロー中タブとフォロワータブの中のユーザー名のリンクからユーザー詳細画面に跳ぶように修正。

Osk0830 reviewed (approved):
お疲れ様でした！
こちらOKなのでマージしますね！
